proof/clear_cofactor_p2.saw

```diff
@@ -0,0 +1,190 @@
+/*
+ * Copyright (c) 2021 Galois, Inc.
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+*/
+
+import "../spec/HashToCurveE2Aux.cry";
+import "../spec/implementation/HashToG2.cry";
+
+// Admitted theorem that doubling a point in projective coordinates has the
+// right effect in affine coordinates.
+point_double_thm <- admit_cryptol
+  {{ \P -> affinify E' (point_double Fp_2 P) == double E' (affinify E' P) }};
+
+psi_psi_double_equiv_thm <- custom_prove_cryptol
+  {{ \P -> affinify E' (psi_psi_double_impl P) == psi_psi_double (affinify E' P) }}
+  do {
+    unfolding ["psi_psi_double", "psi_psi_double_impl"];
+    simplify (addsimps [psi_equiv_thm', point_double_thm, affinify_E'_thm] (cryptol_ss()));
+    w4_unint_z3 ["affinify_E'", "E'"];
+  };
+
+// Prove times_minus_z cryptol spec matches the extracted
+// POINTonE2_times_minus_z term.
+times_minus_z_rev_thm <- custom_prove_cryptol
+  {{ \P -> times_minus_z P == POINTonE2_abs (POINTonE2_times_minus_z (POINTonE2_rep P)) }}
+  do {
+    unfolding ["POINTonE2_times_minus_z"];
+    simplify (addsimps (foldr concat [POINTonE2_thms, e2_curve_op_thms, e2_scalar_mult_alg_thms] []) empty_ss);
+    w4_unint_z3 ["affinify_E'", "E'", "POINTonE2_abs", "e2_scalar_mult"];
+  };
+
+// Prove times_minus_z is equivalent to e2_scalar_mult
+times_minus_z_thm <- custom_prove_cryptol
+  {{ \P -> affinify E' (times_minus_z P) == e2_scalar_mult pos_c1 (affinify E' P) }}
+  do {
+    simplify (addsimps [times_minus_z_rev_thm, affinify_POINTonE2_times_minus_z_thm, affinify_E'_thm] empty_ss);
+    simplify (addsimps POINTonE2_thms empty_ss);
+    w4_unint_z3 ["e2_scalar_mult", "affinify_E'"];
+  };
+
+// Algebraic theorems
+sub_to_add_neg_p2_thm <- admit_cryptol
+  {{ \p1 p2 -> sub E' p1 p2 == add E' p1 (neg E' p2) }};
+
+psi_neg_thm <- prove_cryptol {{ \P -> psi (neg E' P) == neg E' (psi P) }} [];
+
+neg_neg_p2_thm <- prove_cryptol
+  {{ \P -> neg E' (neg E' P) == (P.0, Fp_2.normalize P.1)}} [];
+
+mult_commutes_p2_thm <- admit_cryptol
+  {{ \x y z -> e2_scalar_mult x (add E' y z) == add E' (e2_scalar_mult x y) (e2_scalar_mult x z) }};
+
+pull_mult_neg_p2_thm <- admit_cryptol
+  {{ \x y -> e2_scalar_mult x (neg E' y) == neg E' (e2_scalar_mult x y) }};
+
+push_neg_add_p2_thm <- admit_cryptol
+  {{ \x y -> neg E' (add E' x y) == add E' (neg E' x) (neg E' y) }};
+
+add_normalize_p2_thm <- admit_cryptol
+  {{ \P1 P2 -> add_E' (P1.0, Fp_2.normalize P1.1) P2 == add_E' P1 P2 }};
```

Comment:
I based this theorem on this theorem from the `miller-loop` branch: https://github.com/GaloisInc/BLST-Verification/blob/90848304f3775de04d065d1dfbcc6a20b0920fbb/proof/algebraic-laws.saw#L85

I'm specifically calling it out because I'd like to double check that the same principle applies here too.

Reply:
I do not think this is true.  If P2 == point_O, then `add E' P1 p2` equals `P1`, so if P1 is not normalized this rule is incorrect.

Reply:
Is there something similar to `POINTonE2_invariant` and `POINTonE2_affine_invariant` for points of type `ProjectivePoint t_Fp_2` and `AffinePoint t_Fp_2` respectively?  In [this commit](https://github.com/GaloisInc/BLST-Verification/pull/51/commits/ff5e94ab7cad2a39424ff00024698e718742049d) I removed `add_normalize_p2_thm` and conditioned the proof on an undefined `ProjectivePoint_Fp_2_invariant`.  I also made use of an undefined `POINTonE2_affine_invariant` for some theorems.  I initially tried conditioning the proof on `POINTonE2_invariant (POINTonE2_rep P)` but this is always `True`.  The  important properties of these invariants would be:

1.  `forall P : AffinePoint_Fp_2_invariant P -> ((P.0, Fp_2.normalize P.1) == P)`
2. `forall P : AffinePoint_Fp_2_invariant P -> AffinePoint_Fp_2_invariant (-z * P)`
3. `forall P : AffinePoint_Fp_2_invariant (affinify_E' P) == ProjectivePoint_Fp_2_invariant P`

Reply:
> Is there something similar to POINTonE2_invariant and POINTonE2_affine_invariant for points of type ProjectivePoint t_Fp_2 and AffinePoint t_Fp_2 respectively?

I'm not sure this is even the right equivalence to ask about, hence the further explanation.

Reply:
Property 3 might be the tricky one.  The most obvious invariant properties for these types would be to assert that all of the coordinates satisfy the field invariant, `fp2_invariant`, which ought to give properties 1 and 2, but then I think affinify would always give something satisfying the invariant.   I suppose if we throw in "on_curve" then 3 might hold as well.  So maybe that's it: use `is_point_affine` and `is_point_projective`.